### ISSUE TITLE
Fatal error: Unexpectedly found nil while unwrapping an Optional value: file ComposableArchitecture/ForEachStore.swift, line 143

### ISSUE THREAD

**Issue by asplundo** [pointfreeco/swift-composable-architecture#434]
**Describe the bug**
When filtering a ForEachStore we sometimes get a Fatal error: Unexpectedly found nil while unwrapping an Optional value: file ComposableArchitecture/ForEachStore.swift, line 143
The bug seems to appear in the Todos example as well.
We did not see this behavior before the ForEachStore-fixes.

**To Reproduce**
Tested on device iPhone 7 and simulator iPhone 11:
In the Todos project:
Add 20 Todos
mark 2 of them completed
press the Active segment and directly after the Completed segment.

**Expected behavior**
No Fatal error

**Screenshots**

https://user-images.githubusercontent.com/23285231/111308354-dd206b00-865a-11eb-8597-070421171394.mov

**Environment**
 - Xcode 12.4
 - OS (if applicable): iOS 14.4.1

**Comment by Sajjon** [comment]
I get this too, using Example add "Todos", here are reproduction steps (100% of the time):

```sh
received action:
  AppAction.addTodoButtonTapped

received action:
  AppAction.todo(id:, action: .textFieldChanged)

Fatal error: Unexpectedly found nil while unwrapping an Optional value: file ComposableArchitecture/ForEachStore.swift, line 143
2021-04-03 07:45:54.603460+0200 Todos[35003:6441170] Fatal error: Unexpectedly found nil while unwrapping an Optional value: file ComposableArchitecture/ForEachStore.swift, line 143
```

(I use TCA commit: 1ce105f8bf7457d94385056a1946c6dc6f338556)

**Comment by fabstu** [comment]
In my app I have something similar, but which is probably the result of using ScrollView together with two `ForEachStore` on the same data. Probably the update issue? I'm working around it with a `.deleted` flag on each data and excluding them on rendering, which sometimes looks ugly though and sometimes okay.

**Comment by stephencelis** [comment]
Thanks @fabstu for the report and PR. I've opened up an alternate fix in #471 that tries to contain it in the view layer. Can anyone experiencing this issue give that branch a try and see if it works for them? If so we'll merge it soon :smile:
